### PR TITLE
Improve path-related type hints for `setuptools.Extension()` and `distutils.CCompiler()`

### PR DESCRIPTION
**Pull request by agriyakhetarpal**
## Description

This PR is coming via https://github.com/python/mypy/issues/18107#issuecomment-2459028882 post #12928. I think this should maintain compatibility since Sequence is covariant. In particular, the type annotation of the `sources` parameter in `setuptools.Extension.__init__` has been from `list[StrPath]` to ~~`Sequence[StrPath]`~~ `list[str] | list[StrPath]`. `setuptools.Extension` shouldn't need a mutable sequence.

I also added a few test cases, but this is one of my first pull requests to `typeshed` and towards a typing-related project in general (plus I'm not well-versed with type theory) and I'm unsure if these changes are correct – please feel free to guide me towards making these changes in the correct manner, push changes to my branch directly, or close the PR in case it ends up not being needed. Thank you!

Closes python/mypy#18107

## Additional context

xref: https://github.com/python/mypy/issues/18107, https://github.com/pyodide/pyodide/issues/5160

### CODE REVIEW

**Comment by agriyakhetarpal** [comment]
I see, that gets us back to the original issue.

**Comment by agriyakhetarpal** [comment]
I guess I need to change `CCompiler.compile()`'s stubs, too, but I don't know how invasive that change will be.

File: stubs/setuptools/@tests/test_cases/check_distutils.py
Comment (Avasam) [inline stubs/setuptools/@tests/test_cases/check_distutils.py:18]: I feel like this shows a current issue where you can't have a list of `Path`

File: stubs/setuptools/@tests/test_cases/check_extension.py
Comment (Avasam) [inline stubs/setuptools/@tests/test_cases/check_extension.py:8]: `setuptools.Extensions.__init__` still has `sources` as `list[StrPath]` in your branch, this shows that passing inline lists, the type-checkers are able to infer all these as being `list[StrPath]`

Here's a small snippet to demonstrate:
```py
from __future__ import annotations
from os import PathLike
from pathlib import Path


def loose(foo: list[str] | list[PathLike[str]] | list[Path] | list[StrPath]): ...
def str_or_strpath(foo: list[str] | list[StrPath]): ...
def restrictive(foo: list[StrPath]): ...


def _(a: list[str], b: list[PathLike[str]], c: list[Path], d: list[StrPath]) -> None:
    loose(a)
    loose(b)
    loose(c)
    loose(d)
    restrictive(a) # Argument 1 to "restrictive" has incompatible type "list[str]"; expected "list[Union[str, PathLike[str]]]"
    restrictive(b) # Argument 1 to "restrictive" has incompatible type "list[PathLike[str]]"; expected "list[Union[str, PathLike[str]]]"
    restrictive(c) # Argument 1 to "restrictive" has incompatible type "list[Path]"; expected "list[Union[str, PathLike[str]]]"
    restrictive(d)
    restrictive(["file1.c", "file2.c"])  # list of str(s)
    restrictive([Path("file1.c"), Path("file2.c")])  # list of Path(s)
    restrictive([Path("file1.c"), "file2.c"])  # mixed str(s) and Path(s)
    str_or_strpath(a)
    str_or_strpath(b) # Argument 1 to "str_or_strpath" has incompatible type "list[PathLike[str]]"; expected "Union[list[str], list[Union[str, PathLike[str]]]]"
    str_or_strpath(c) # Argument 1 to "str_or_strpath" has incompatible type "list[Path]"; expected "Union[list[str], list[Union[str, PathLike[str]]]]"
    str_or_strpath(d)
    str_or_strpath(["file1.c", "file2.c"])  # list of str(s)
    str_or_strpath([Path("file1.c"), Path("file2.c")])  # list of Path(s) # Argument 1 to "str_or_strpath" has incompatible type "list[Path]"; expected "Union[list[str], list[Union[str, PathLike[str]]]]"
    str_or_strpath([Path("file1.c"), "file2.c"])  # mixed str(s) and Path(s) # Argument 1 to "str_or_strpath" has incompatible type "list[object]"; expected "Union[list[str], list[Union[str, PathLike[str]]]]"

```

**Review by Avasam — CHANGES_REQUESTED** [review]
list's invariance really bites us here. But I think going with generics should solve all issues (quick sanity check: 
image

When appropriate, could you also keep the same methods in setuptools and stdlib's distutils in sync ?
- `Extension` class in setuptools
- `CCompoiler.compile` method in distutils

I'm working towards syncing `setuptools._distutils` no longer being necessary https://github.com/pypa/setuptools/issues/4689

File: stubs/setuptools/setuptools/_distutils/extension.pyi
Comment (agriyakhetarpal) [reply stubs/setuptools/setuptools/_distutils/extension.pyi:23]: Yes, if `Extension.sources` retrieves a list anyway, requiring a list in the input seems like it's just an unnecessary restriction. The long-term solution with generics is interesting and makes (somewhat) sense to me, though – would you like to see me try a follow-up PR to pypa/distutils with that approach sometime soon?

Comment (Avasam) [reply stubs/setuptools/setuptools/_distutils/extension.pyi]: I don't think any change to pypa/distutils is necessary to make this class generic in stubs.

Whilst it's currently not pretty, I'd be fine doing it as a follow-up and keeping this PR as-is to restore functionality to projects that were relying on the previous annotations.

I'm curious as to what other maintainers think though. Is the current state of this PR acceptable as a stop-gap with a follow-up? Or should we make it generic now? Other thoughts? (CCing @hauntsaninja from conversation in #18107 )

Comment (JelleZijlstra) [reply stubs/setuptools/setuptools/_distutils/extension.pyi]: I'd personally prefer to avoid the generic. Generics are confusing for users who use the class in a type annotation, because they need to figure out what the generic parameter refers to. That's easy for something where the generic nature is really inherent to the class (like with `list` or other collections), but for this class, the generic parameter probably doesn't make a huge difference to most users. It may be better to go with a more general type like Sequence to improve usability.

Comment (JelleZijlstra) [reply stubs/setuptools/setuptools/_distutils/extension.pyi]: Also I wrote this before I saw @Avasam's comment above. I think the current state of this PR is OK; the one thing that gives me pause is the `list[PathLike[str]] | list[str] ...` annotation. Type checkers aren't terribly likely to infer `list[PathLike[str]]` as the type for a list that the user provides, so that type isn't very usable in practice. We could consider using `Sequence` instead, but then we'd get false negatives if a user passes a tuple or other sequence. Ideally we'd solve this in the runtime of setuptools so it accepts any Sequence and not just list, but in the meantime what the PR currently does seems like a good solution.

**Comment by agriyakhetarpal** [comment]
Please let me know if there's anything more I can or need to address here. Thanks for the review so far!

**Comment by agriyakhetarpal** [comment]
Question: should we just use `Sequence[StrPath]` like we tried before to make CPython happy?

**Comment by Avasam** [comment]
> Question: should we just use `Sequence[StrPath]` like we tried before to make CPython happy?

For the method argument? I think Jelle's answer sums it up well:
> We could consider using `Sequence` instead, but then we'd get false negatives if a user passes a tuple or other sequence. Ideally we'd solve this in the runtime of setuptools so it accepts any Sequence and not just list, but in the meantime what the PR currently does seems like a good solution.

So I'd leave it as-is, open a PR at https://github.com/pypa/distutils to change the runtime check to any `Iterable` (non-`str`, the "not str" check is necessary because a `str` is a valid `Iterable[str]` and `Sequence[str]`, see https://github.com/python/mypy/issues/11001)

**Review by Avasam — APPROVED** [review]

**Comment by agriyakhetarpal** [comment]
Thanks for the feedback. I opened a PR in pypa/distutils here: https://github.com/pypa/distutils/pull/311 which allows this behaviour.

**Comment by Avasam** [comment]
Thanks for your contribution!

This fixes an issue for mypy itself and Jelle seems to approve of the change as a stop-gap measure until the situation is improved upstream: https://github.com/pypa/distutils/pull/311
> in the meantime what the PR currently does seems like a good solution.

So I'll merge and if any other issue come up feel free to open a follow-up PR 😃

**Pull request merged**
